source/clustering.md

```diff
@@ -374,37 +444,121 @@ we sum them together to get the *total WSSD*.
+        alt.X("flipper_length_standardized"),
+        alt.Y("bill_length_standardized"),
+        alt.Color('label:N').legend(None),
+        alt.Shape('label:N').legend(None)
+)
+
+glue('toy-kmeans-init-1', points_kmeans_init, display=True)
+```
+
 We begin the K-means algorithm by picking K,
 and randomly assigning a roughly equal number of observations
 to each of the K clusters.
-An example random initialization is shown in {numref}`toy-kmeans-init-1`.
+An example random initialization is shown in {numref}`toy-kmeans-init-1`
 
 
+:::{glue:figure} toy-kmeans-init-1
+:figwidth: 700px
+:name: toy-kmeans-init-1
 
-```{figure} img/clustering/toy-kmeans-init-1.png
----
-height: 400px
-name: toy-kmeans-init-1
----
 Random initialization of labels.
+:::
+
+```{code-cell} ipython3
+:tags: [remove-cell]
+
+from sklearn.cluster import KMeans
 
+
+def plot_kmean_iterations(iterations, data, centroid_init):
+    """Plot kmeans cluster and label updates for multiple iterations"""
+    dfs = []
+    centroid_inits = []
+    for i in range(1, iterations+1):
+        data['iteration'] = f'Iteration {i}'
+        data['update_type'] = 'Center Update'
+        data['flipper_centroid'] = data['label'].map(centroid_init['flipper_length_standardized'])
+        data['bill_centroid'] = data['label'].map(centroid_init['bill_length_standardized'])
+        dfs.append(data.copy())
+
+        cluster_columns = ['flipper_length_standardized', 'bill_length_standardized']
+        knn = KMeans(init=centroid_init[cluster_columns], n_clusters=3, max_iter=1, n_init=1)
+        knn.fit(data[cluster_columns])
+
+        data['label'] = knn.labels_
+        data['iteration'] = f'Iteration {i}'
+        data['update_type'] = 'Label Update'
+        data['flipper_centroid'] = data['label'].map(centroid_init['flipper_length_standardized'])
+        data['bill_centroid'] = data['label'].map(centroid_init['bill_length_standardized'])
+        dfs.append(data.copy())
+
+        centroid_init = data.groupby('label').mean(numeric_only=True)
+
+    points = alt.Chart(
+        pd.concat(dfs),
+        width=200,
+        height=200
+    ).mark_point(filled=True, size=75, opacity=1).encode(
+        alt.X("flipper_length_standardized").scale(domain=(-2, 2)),
+        alt.Y("bill_length_standardized").scale(domain=(-2, 2)),
+        alt.Color('label:N').legend(None),
+        alt.Shape('label:N').legend(None)
+    )
+
+    centroids = points.mark_point(size=200, filled=True, stroke='black', strokeWidth=1.5).encode(
+        alt.X("mean(flipper_centroid)")
+            .scale(domain=(-2, 2))
+            .title("Flipper Length (standardized)"),
+        alt.Y("mean(bill_centroid)")
+            .scale(domain=(-2, 2))
+            .title("Flipper Length (standardized)")
+    )
+
+    return (points + centroids).facet(
+        row=alt.Row('iteration', header=alt.Header(title='', labelFontSize=18)),
+        column=alt.Column('update_type', header=alt.Header(title='', labelFontSize=18))
```

Comment:
I changed the default layout here to be only in two columns, one for each condition. I think this is much clearer than alternating between the two conditions in four columns, so ideally the same change would be made in the R book, but I wont be able to make it in the near future myself.

I also reduced the number of iterations with 1 since it is already clear what is happening at that point

Reply:
We did 2 rows here because the single-column format did not look good in the print book

(if you read the text right before the image, it's actually wrong in the R book -- it's the old description where we originally had something more like what you have now -- I've now filed an issue in the R book to fix that text.)

BUT I agree that we can remove the last iteration, in which case maybe it would look fine (and in that case the 2 column format is better).

Let's keep what you have here and I'll change the R version.